code/poolams.c

```diff
@@ -756,6 +756,7 @@ static void AMSVarargs(ArgStruct args[MPS_ARGS_MAX], va_list varargs)
   args[2].key = MPS_KEY_AMS_SUPPORT_AMBIGUOUS;
   args[2].val.b = va_arg(varargs, Bool);
   args[3].key = MPS_KEY_ARGS_END;
+  AVER(MPS_ARGS_MAX - 1 > 3);
```

Comment:
1. I think it would make sense to put the assertion first, before dereferencing `args`, otherwise if `args` is too short we get the undefined behaviour before the assertion that is supposed to prevent it.
2. The assertion seems over-strict as the condition we need is `MPS_ARGS_MAX - 1 >= 3`.

Reply:
1. The idea is to put the assertion next to the line that would need to change with it.  The assertion is static at compile-time, so there's no "before" as such.
2. I'm pretty sure filling in args[0] to args[3] uses four slots.  Have I made an obi-wan error?

Reply:
I'm not sure what you mean by "The assertion is static at compile-time". If I change `3` to `999` (say) then I get the assertion failure at runtime, not at compile time
```
$ make -f xca6ll.gmk VARIETY=cool amssshe
xca6ll: xca6ll/cool/amssshe
$ xca6ll/cool/amssshe 
xca6ll/cool/amssshe: randomize(): choosing initial state (v3): 1673032430.
poolams.c:759: MPS ASSERTION FAILED: MPS_ARGS_MAX - 1 > 999

Abort trap: 6
```

Reply:
I don't think we have static assertions in the MPS, but if we need them we should add them.

Reply:
> I'm not sure what you mean by "The assertion is static at compile-time".

I mean that the outcome is known to the compiler at compile-time.  We have lots of assertions like this.  See, for example, MPMCheck https://github.com/Ravenbrook/mps/blob/5051593b21c80f12d30437ed4eccf197988db98c/code/mpm.c#L37-L98

Reply:
My point is that in the scenario where this assertion would fire, the code is incorrect at compile time, and the compiler "knows" it.  There are no run-time conditions that *won't* fire the assertion. There's a very slim chance that writing to the arg will cause, e.g. a fault before the assertion fires, but I think there's more of a chance of a coding error by separating the two lines.  That's what I was thinking.

We could write things differently: introduce more symbols, remove the numbers, etc. but it's getting a bit elaborate.

Reply:
On balance I prefer keeping the AVER near to the line that would cause it to change